src/RileyMaze/userFunction.js

```diff
@@ -0,0 +1,182 @@
+            this._model.disconnect(this._notifyHandler);
+            this._notifyHandler = null;
+        }
+
+        this._model = null;
+    }
+
+    _bindModel(model) {
+        if (this._model)
+            this._unbindLevelModel();
+
+        this._model = model;
+        this._notifyHandler = model.connect('notify', this._setCode.bind(this));
+        this._setCode();
+    }
+
+    _setCode() {
+        const {name, args, perLevel, modelProp} = USER_FUNCTIONS[this._codeview.userFunction];
+
+        let code;
+        if (perLevel)
+            code = this._model[modelProp];
+        else
+            code = this._global[modelProp];
+
+        this._codeview.text = `function ${name}(${args.join(', ')}) {
+${code}
+}`;
+    }
+
+    _updateCode(funcBody) {
+        const {perLevel, modelProp} = USER_FUNCTIONS[this._codeview.userFunction];
+        // Block the normal notify handler that updates the code view, since we
+        // are propagating updates from the codeview to the GUI.
+        if (perLevel) {
+            GObject.signal_handler_block(this._model, this._notifyHandler);
+            try {
+                this._model[modelProp] = funcBody;
+            } finally {
+                GObject.signal_handler_unblock(this._model, this._notifyHandler);
+            }
+        } else {
+            GObject.signal_handler_block(this._global, this._globalNotifyHandler);
+            try {
+                this._global[modelProp] = funcBody;
+            } finally {
+                GObject.signal_handler_unblock(this._global, this._globalNotifyHandler);
+            }
+        }
+    }
+
+    _compile() {
+        const {name, args, getScope} = USER_FUNCTIONS[this._codeview.userFunction];
+        const code = this._codeview.text;
+        const scope = getScope();
+        let userFunction;
+
+        try {
+            // eslint-disable-next-line no-new-func
+            const factoryFunc = new Function('scope', `\
+with(scope){
+${code}
+; if (typeof ${name} !== 'undefined') return ${name};
+}`);
+            userFunction = factoryFunc(scope);
+        } catch (e) {
+            if (!(e instanceof SyntaxError || e instanceof ReferenceError))
+                throw e;
+            this._codeview.setCompileResultsFromException(e);
+            return;
+        }
+
+        if (!userFunction) {
+            // Function definition deleted, or otherwise not working
+            this._codeview.setCompileResults([{
+                start: {
+                    line: 1,
+                    column: 1,
+                },
+                end: {
+                    line: 1,
+                    column: 1,
+                },
+                message: `Expected a function named ${name}.`,
+                fixme: `function ${name}(${args.join('\n')}) {\n}\n`,
+            }]);
+            return;
+        }
+
+        try {
+            userFunction();
+        } catch (e) {
+            this._codeview.setCompileResultsFromException(e);
```

Comment:
In a few cases, the instructions codeview should work differently from all the other codeviews and call `_updateCode` with erroneous code, in order to pass it back to the game. This happens in cases where the game should display a glitched instruction tile. Double check with Randy when this is supposed to happen. My understanding is — typo in instruction name; too many instructions; too few instructions. But maybe he wants to have glitches in the game whenever there is an error at all, like in the case of invalid JS code?